src/crimson/osd/ops_executer.h

```diff
@@ -275,4 +265,23 @@ OpsExecuter::osd_op_errorator::future<> OpsExecuter::flush_changes(
   }
 }
 
+class PgOpsExecuter {
+  PG& pg;
+  Ref<MOSDOp> msg;
+
+  // PG operations are being provided with pg instead of os.
+  template <class Func>
+  auto do_pg_op(Func&& f) {
+    return std::forward<Func>(f)(std::as_const(pg),
+                                 std::as_const(msg->get_hobj().nspace));
+  }
+
+public:
+  PgOpsExecuter(PG& pg, Ref<MOSDOp> msg)
+    : pg(pg), msg(std::move(msg)) {
+  }
+
+  seastar::future<> execute_pg_op(class OSDOp& osd_op);
```

Comment:
```suggestion
  seastar::future<> execute_op(class OSDOp& osd_op);
```

as the name of the class implies that the ops being executed are PG Ops.

Reply:
for the same reason, might want to s/execute_osd_op/execute_op/ in `OpsExecuter`?

Reply:
also, might want to add a comment on top of `OpsExecuter` to explain that that class is solely for executing ops targeting objects. and add a comment on top of `PgOpsExecuter` to explain that that class is solely for executing ops targeting a certain PG.

Reply:
> as the name of the class implies that the ops being executed are PG Ops.

Agreed, the duplication is a downside; the upside is easier grepping. No strong feelings here.

Reply:
I'm for the easy grep. Analyzing current Ceph code is a slow process (without a good and fast IDE), in part due to this issue.